### ISSUE TITLE
Remove double loop in J

### ISSUE THREAD

**Issue by davidweichiang** [diprism/fggs#168]
It can be changed into a single loop (thanks @darcey).

**Comment by ccshan** [comment]
Can I get more explanation?

**Comment by davidweichiang** [comment]
Currently J uses a double loop to compute $\prod_{i \in [n] \setminus \{k\}} weight(e_i)$ where $weight(e_i)$ is the sum-product of $e_i$'s label if nonterminal, or the weight of its factor if terminal. So it runs in O(|E|^2) time. But it can be reduced to O(|E|) time as follows:

- Number the edges $e_1, ..., e_n$
- For k = 0 to n, compute $\prod_{i=1,...,k} weight(e_i)$ in constant time
- For k = n downto 0, compute $\prod_{i=k,...,n} weight(e_i)$ in constant time
- For k = 1 to n, compute $\prod_{i \in [n] \setminus \{k\}} weight(e_i)$ in constant time

**Comment by davidweichiang** [comment]
Unfortunately this isn't the easiest fix, because the outer loop is in `J` (and `J_log`) but the inner loop is in `sum_product_edges`.

**Comment by davidweichiang** [comment]
Right. I guess the ordering does matter, which is something we haven't thought about at all yet. People who work on tensor networks have thought about this a lot.

**Comment by davidweichiang** [comment]
@grantexley is starting to work on this. But one possible snag we thought of is that we will not be able to make one big call to einsum, but many smaller calls. But semiring-einsum has some optimizations for 3+ arguments, and we will likely break those optimizations. We're not sure yet if that's bad.

**Comment by davidweichiang** [comment]
Follow-up based on meeting with @grantexley today: The comment above says "keeping all nodes that appear in any $e_i$ where $i > k$ (as well as the original `rule.rhs.ext`)", which we are calling "future nodes" (when computing the prefix products of the $e_i$). To clarify, this should be "keeping all nodes $v$ such that $v$ is an endpoint of any $e_i$ where $i\le k$ and $v$ is an endpoint of any $e_i$ where $i>k$ (as well as the original `rule.rhs.ext`)." (Maybe @ccshan considered the first condition to be already implicit in the word "keeping".)

**Comment by grantexley** [comment]
<img width="738" alt="image" src="https://github.com/user-attachments/assets/d3098812-7d01-44dd-af94-5dc69fc866f8" />
<img width="661" alt="image" src="https://github.com/user-attachments/assets/ff722463-80c2-40b4-a77a-d57d39b10612" />
